### ISSUE TITLE
Training with Ray Clusters Ultralytics Model

### ISSUE THREAD

**Issue by athava-sthenos** [ultralytics/ultralytics#24273]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

Is there documentation about how to implement distributed training with ray clusters?

Being on a single node with multiple GPUs, one executes ultralytics.YOLO.train(). Then ultralytics is responsible for training the YOLO model on the cuda visible devices.

But, how could YOLO training be done when multiple nodes with multiple workers are provided by a ray cluster? Is there documentation for such a case?

Thank you

**Comment by UltralyticsAssistant** [comment]
👋 Hello @athava-sthenos, thank you for your interest in Ultralytics 🚀! This is an automated response to help get things moving quickly, and an Ultralytics engineer will also assist you soon.
We recommend a visit to the [Docs](https://docs.ultralytics.com/) for new users where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples and where many of the most common questions may already be answered.
If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug it.
If this is a custom training ❓ Question, please provide as much information as possible, including your cluster setup, Ray worker configuration, dataset details, training command or code snippet, logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).
Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.
## Upgrade
Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:
```bash
pip install -U ultralytics
```
## Environments
YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):
- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo26"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>
## Status
<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>
If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

**Comment by Y-T-G** [comment]
Are you trying to run hyparameter tuning?

**Comment by athava-sthenos** [comment]
> Are you trying to run hyparameter tuning?

no, just train model with distributed training on ray cluster

**Comment by Y-T-G** [comment]
You can try this:

```python
"""
Multi-node DDP training for Ultralytics YOLO, launched via Ray Train.

Launch from any machine that can reach the Ray cluster:

    python ray_yolo_train.py --model yolo11n.pt --data coco8.yaml \
        --nodes 2 --gpus-per-node 4

Or submit as a Ray job (no ssh loop needed):

    ray job submit --address=http://<head-node>:8265 --working-dir . \
        -- python ray_yolo_train.py --nodes 2 --gpus-per-node 4

Prereqs:
  * A Ray cluster already running:
      head:   ray start --head --port=6379 --dashboard-host=0.0.0.0
      worker: ray start --address=<head-ip>:6379
    (Or KubeRay, or the Ray cluster launcher.)
  * Shared filesystem reachable at --storage-path on every node; dataset
    paths inside --data must also resolve on every node.
  * Identical Ultralytics / PyTorch / CUDA / NCCL versions cluster-wide.
"""

import argparse
import os

import ray
import torch
import torch.distributed as dist
from ray.train import Checkpoint, RunConfig, ScalingConfig, get_context
from ray.train.torch import TorchTrainer


def train_func(config: dict):
    """Runs once on every Ray worker (one process per GPU across all nodes).

    Ray Train has already:
      - launched this process with the right CUDA_VISIBLE_DEVICES,
      - set MASTER_ADDR, MASTER_PORT, RANK, WORLD_SIZE, LOCAL_RANK,
      - initialized a torch.distributed process group.

    Ultralytics' BaseTrainer._setup_ddp() wants to init the process group
    itself. We tear Ray's down first so Ultralytics can re-init cleanly
    against the same MASTER_ADDR/PORT that Ray already set up.
    """
    # Lazy import so Ray's runtime_env / worker image doesn't need it at submit time.
    from ultralytics import YOLO

    ctx = get_context()
    rank = ctx.get_world_rank()
    local_rank = ctx.get_local_rank()
    world_size = ctx.get_world_size()

    # Hand the process group back to Ultralytics.
    if dist.is_initialized():
        dist.destroy_process_group()

    # Ultralytics reads these env vars to decide it's in DDP-worker mode
    # and to skip its usual torch.distributed.run subprocess spawn.
    os.environ["RANK"] = str(rank)
    os.environ["LOCAL_RANK"] = str(local_rank)
    os.environ["WORLD_SIZE"] = str(world_size)

    torch.cuda.set_device(local_rank)

    model = YOLO(config["model"])

    # Because LOCAL_RANK is set, BaseTrainer.train() runs _do_train() directly
    # in this process instead of spawning torchrun children. _setup_ddp() will
    # call dist.init_process_group() using the MASTER_ADDR/PORT Ray set.
    model.train(
        data=config["data"],
        device=local_rank,          # this worker owns exactly one GPU
        **config["train_args"],
    )

    # Only rank 0 reports results + checkpoint back to Ray Train.
    if rank == 0 and model.trainer is not None:
        best_path = str(model.trainer.best)          # e.g. .../weights/best.pt
        ckpt_dir = os.path.dirname(best_path)
        ckpt = Checkpoint.from_directory(ckpt_dir) if os.path.isdir(ckpt_dir) else None
        ray.train.report(
            metrics={"fitness": float(model.trainer.fitness or 0.0)},
            checkpoint=ckpt,
        )


def main():
    p = argparse.ArgumentParser()
    p.add_argument("--model", default="yolo11n.pt")
    p.add_argument("--data", default="coco8.yaml")
    p.add_argument("--epochs", type=int, default=100)
    p.add_argument("--imgsz", type=int, default=640)
    p.add_argument("--batch", type=int, default=64,
                   help="GLOBAL batch size; Ultralytics splits across ranks.")
    p.add_argument("--nodes", type=int, default=2)
    p.add_argument("--gpus-per-node", type=int, default=4)
    p.add_argument("--cpus-per-worker", type=int, default=8)
    p.add_argument("--storage-path", default="/mnt/shared/ray_results",
                   help="Shared FS path every node can write to.")
    p.add_argument("--ray-address", default="auto",
                   help="'auto' to attach to a local cluster, or ray://host:10001")
    args = p.parse_args()

    ray.init(address=args.ray_address)

    num_workers = args.nodes * args.gpus_per_node

    trainer = TorchTrainer(
        train_loop_per_worker=train_func,
        train_loop_config={
            "model": args.model,
            "data": args.data,
            "train_args": {
                "epochs": args.epochs,
                "imgsz": args.imgsz,
                "batch": args.batch,
                "project": "ray_yolo",
                "name": "run",
                "exist_ok": True,
            },
        },
        scaling_config=ScalingConfig(
            num_workers=num_workers,
            use_gpu=True,
            resources_per_worker={"GPU": 1, "CPU": args.cpus_per_worker},
        ),
        run_config=RunConfig(
            name="yolo_multinode_ddp",
            storage_path=args.storage_path,
        ),
    )

    result = trainer.fit()
    print(f"Finished. Metrics: {result.metrics}")
    print(f"Best checkpoint: {result.checkpoint}")


if __name__ == "__main__":
    main()
```
